File: 0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        for(int i = 0; i < matrix.size(); i++){
            
            int first = matrix[i][0]; 
            int last = matrix[i][matrix[i].size()-1]; 

            //vector<int> curr = matrix[i]; 
            
            if(first <= target && last >= target){
                int start = 0; 
                int end = matrix[i].size()-1; 

                while(start <= end){
                    int mid = (end + start) / 2;  

                    if(matrix[i][mid] == target){
                        return true; 
                    } else if(matrix[i][mid] < target){
                        start = mid + 1; 
                    } else{
                        end = mid - 1; 
                    }
                }
            }
        }

        return false; 
    }
};

File: 0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length; 
        for(int i = 0; i < m; i++){
            int[] candidate = matrix[i]; 
            if(candidate[0] <= target && candidate[candidate.length-1] >= target){
                int left = 0, right = candidate.length-1; 
                while(left <= right){
                    int mid = left + (right - left) / 2; 
                    if(candidate[mid] == target) return true; 

                    if(candidate[mid] > target) right = mid - 1;
                    else left = mid + 1; 
                }
            }
        }

        return false; 
    }
}